### ISSUE TITLE
What values should normal ACI modules return

### ISSUE THREAD

**Issue by dagwieers** [datacenter/aci-ansible#107]
So we deliberately postponed this discussion, but now the time to make that decision is nearing. What return values should ACI modules have (and we are now excluding aci_rest from this discussion).

There are three different configurations:
1. The proposed configuration (so the datstructure we send to ACI)
2. The existing configuration (what was there before)
3. The new running configuration (what is there after a change)

There are a few concerns:
1. Providing more than one dataset can/will lead to end-user confusion (we're talking about a vast amount of data that is mostly similar). End-user confusion leads to difficult support situations. @dagwieers
2. A single dataset makes it harder to confirm changes when troubleshooting, and makes it harder to test for expected behaviour

Possible solutions:
- Add a generic parameter to switch on/off the desired output level (other modules do something similar) @dagwieers

*Feel free to edit this ticket*

**Comment by jmcgill298** [comment]
I think we should have a ranking of value/importance for end users and supply data based on verbosity level.

My rankings would go like this:
no verbosity: ok/changed/fail
1 verbosity: add existing and proposed
>I would say that proposed can be derived from the input parameters, and final state can be derived from existing and proposed. This is a Rest API, so it POST and DELETE should be all or nothing. If something breaks due to playbook execution, I am concerned with what was there before (existing) and what changed (proposed). Final state would include many values that are of no concern, thus making it unclear what needs to be reverted.
2 verbosity: add proposed
3 verbosity: add HTTP status code, response, etc.

How would everyone else rank the return data?

**Comment by dagwieers** [comment]
Well, using verbosity to change what the module returns is really frowned upon within Ansible. It can lead to the problem that as soon as people troubleshoot there's something there, that isn't when using. Especially since you can not influence verbosity for a task individually from the playbook itself. So you don't get reproducible states depending on the number of -v flags.

We could make the exception for debug mode, but yet I would prefer using a module parameter, like the `return_content` option in **uri*. At least you can influence it, and it is documented and reproducible. So we could go for a `return_details` or something like that.

**Comment by jmcgill298** [comment]
Ya, that is fine. The main idea is that results are up to the user. If everyone agrees on that, then it is a matter of what options we want to give the users.

**Comment by dagwieers** [comment]
Right, so for me the user is either debugging, or is not debugging. So a `return_details` (or whatever name is deemed more appropriate) is sufficient. But maybe there are other use-cases.

For integration tests, it could be useful to get this information as well, but getting more than you need when troubleshooting, is not a real problem to me. The idea is not to confuse the user by default :-)

**Comment by jmcgill298** [comment]
The bool makes the user input easier, but I think that is what creates the simplicity vs complexity problem. I think my tiered approach is a middle ground between user input vs desired output. The other option is the buffet approach and let users choose each individually as they please (I would place this as my least preferred option).

As far as what to offer for the boolean approach, I expect us to disagree based on our perspectives. Coming from TAC and building labs, the valuable data is really just the end state (same for anyone doing build activities). I come from an Ops background and so I would be more concerned with knowing what changed, and there are tons of diff tools out there for a reason (no one wants to compare data side by side).

Ultimately I am fine with providing only the end state

**Comment by dagwieers** [comment]
So using YAML anchors and references, it is easy to define your preferred ACI options once, and reuse them for every ACI task if you like. This is how I usually do the aci_login parameters. (See the demo playbook in this repository).

But yes, I definitely understand your point of view, but the standard way for Ansible is changed/failed return values, and check-mode and diff output. None of the modules really show the details before and after unless specifically asked with --diff. (And we already support it !)

So if everyone agrees, we will implement a `return_details` which will return everything but is off by default.

Now, what are the return keys we want to use, and should we look at other modules in the network-space to make this convenient for people using heterogeneous envs ?

**Comment by jmcgill298** [comment]
I agree on looking at other modules and seeing what applies here before coming up with our own stuff

**Comment by dagwieers** [comment]
So we really need to get this one closed ASAP.

The modules are tagged as status **preview**, which means that the interfaces (parameters and result values) are not set in stone and may change in the future. So we do have some wiggle room after v2.4 to change this, but I would be very reluctant to do this. In any case if we can prevent it now, we should.

**Comment by dagwieers** [comment]
I made a proposal here: https://github.com/ansible/proposals/issues/93
So we are waiting on a final decision by the core team for the best implementation.
I am going to close this issue and remove the issue tracker from the interface.
It's the first step in dismantling this repository.
